### PR TITLE
Update contract failure handling

### PR DESCRIPTION
**Pull request by LukaszRozmej**

### CODE REVIEW

File: src/Nethermind/Nethermind.Consensus.AuRa/Validators/ReportingContractBasedValidator.PersistentReports.cs [hunk 1]
```diff
@@ -118,7 +119,7 @@ private void FilterReports(BlockHeader parent)
                         if (_logger.IsTrace) _logger.Trace("Successfully removed report from report cache.");
                     }
                 }
-                catch (AuRaException e)
+                catch (AbiException e)
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Consensus.AuRa/Validators/ReportingContractBasedValidator.PersistentReports.cs:122]: in the other file you change AbiException to AuRa and here you changed the exception that you catch from AuRa to Abi. Was it intended?

File: src/Nethermind/Nethermind.Consensus.AuRa/Validators/ReportingContractBasedValidator.PersistentReports.cs [hunk 2]
```diff
@@ -91,7 +92,7 @@ private void ResendPersistedReports(BlockHeader blockHeader)
                         {
                             SendTransaction(ReportType.Malicious, _posdaoTxSender, CreateReportMaliciousTransactionCore(persistentReport));
                         }
-                        catch (AuRaException e)
+                        catch (AbiException e)
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Consensus.AuRa/Validators/ReportingContractBasedValidator.PersistentReports.cs:95]: in the other file you change AbiException to AuRa and here you changed the exception that you catch from AuRa to Abi. Was it intended?

File: src/Nethermind/Nethermind.Consensus.AuRa/Validators/ContractBasedValidator.cs [hunk 3]
```diff
@@ -240,14 +240,21 @@ private void InitiateChange(Block block, Address[] potentialValidators, bool isP
...
+                if (validators.Length == 0)
+                {
+                    throw new AuRaException("Failed to initialize validators list.");
+                }
 
-            if (validators.Length == 0)
+                return validators;
+            }
+            catch (AbiException e)
             {
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Consensus.AuRa/Validators/ContractBasedValidator.cs:255]: you change AbiException to AuRa and  you changed the exception that you catch from AuRa to Abi. Was it intended?

File: src/Nethermind/Nethermind.Consensus.AuRa/Validators/ContractBasedValidator.cs [hunk 4]
```diff
@@ -240,14 +240,21 @@ private void InitiateChange(Block block, Address[] potentialValidators, bool isP
...
         private Address[] LoadValidatorsFromContract(BlockHeader parentHeader)
         {
-            var validators = ValidatorContract.GetValidators(parentHeader);
+            try
+            {
+                var validators = ValidatorContract.GetValidators(parentHeader);
+                
+                if (validators.Length == 0)
+                {
+                    throw new AuRaException("Failed to initialize validators list.");
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Consensus.AuRa/Validators/ContractBasedValidator.cs:249]: you change AbiException to AuRa and  you changed the exception that you catch from AuRa to Abi. Was it intended?

File: src/Nethermind/Nethermind.Consensus.AuRa/Validators/ContractBasedValidator.Posdao.cs [hunk 5]
```diff
@@ -42,7 +43,7 @@ public IEnumerable<Transaction> GetTransactions(BlockHeader parent, long gasLimi
                     {
                         emitInitChangeCallable = ValidatorContract.EmitInitiateChangeCallable(parent);
                     }
-                    catch (AuRaException e)
+                    catch (AbiException e)
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Consensus.AuRa/Validators/ContractBasedValidator.Posdao.cs:46]: you change AbiException to AuRa and  you changed the exception that you catch from AuRa to Abi. Was it intended?

File: src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxPermissionFilter.cs [hunk 6]
```diff
@@ -91,7 +92,7 @@ private ITransactionPermissionContract.TxPermissions GetPermissions(Transaction
                 {
                     (txPermissions, shouldCache) = versionedContract.AllowedTxTypes(parentHeader, tx);
                 }
-                catch (AuRaException e)
+                catch (AbiException e)
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxPermissionFilter.cs:95]: you change AbiException to AuRa and  you changed the exception that you catch from AuRa to Abi. Was it intended?

**Review by tkstanczak — APPROVED** [review]

**Comment by LukaszRozmej** [comment]
Answering questions about AuRaException - yes. I've double checked and  in 2 places it is thrown its directly handled. The only other place is ContractBasedValidator.LoadValidatorsFromContract where it is supposed to bubble up and stop processing chain as the chain is corrupted when we can't load validators from it.

**Pull request merged**